File: src/main/java/xyz/avarel/waffles/nodes/WaffleNode.java
package xyz.avarel.waffles.nodes;

import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.nodes.Node;
import com.oracle.truffle.api.nodes.UnexpectedResultException;
import xyz.avarel.waffles.WaffleFunction;
import xyz.avarel.waffles.WaffleTypesGen;

public abstract class WaffleNode extends Node {
    public abstract Object execute(VirtualFrame virtualFrame);

    public int executeInt(VirtualFrame virtualFrame) throws UnexpectedResultException {
        return WaffleTypesGen.expectInteger(execute(virtualFrame));
    }

    public boolean executeBoolean(VirtualFrame virtualFrame) throws UnexpectedResultException {
        return WaffleTypesGen.expectBoolean(execute(virtualFrame));
    }

    public WaffleFunction executeFunction(VirtualFrame virtualFrame) throws UnexpectedResultException {
        return WaffleTypesGen.expectWaffleFunction(execute(virtualFrame));
    }

    protected boolean isArgumentIndexInRange(VirtualFrame virtualFrame, int index) {
        return index < virtualFrame.getArguments().length - 1;
    }

    protected Object getArgument(VirtualFrame virtualFrame, int index) {
        return virtualFrame.getArguments()[index + 1];
    }
}

File: src/main/kotlin/xyz/avarel/waffles/nodes/WaffleNode.kt
package xyz.avarel.waffles.nodes

import com.oracle.truffle.api.frame.VirtualFrame
import com.oracle.truffle.api.nodes.Node
import xyz.avarel.waffles.WaffleFunction
import xyz.avarel.waffles.WaffleTypesGen

abstract class WaffleNode : Node() {
//    var sourceSection: SourceSection? = null

    abstract fun execute(virtualFrame: VirtualFrame): Any

//    open fun executeVoid(virtualFrame: VirtualFrame) {
//
//    }

    open fun executeInt(virtualFrame: VirtualFrame): Int {
        return WaffleTypesGen.expectInteger(execute(virtualFrame))
    }

    open fun executeString(virtualFrame: VirtualFrame): String {
        return WaffleTypesGen.expectString(execute(virtualFrame))
    }

    open fun executeBoolean(virtualFrame: VirtualFrame): Boolean {
        return WaffleTypesGen.expectBoolean(execute(virtualFrame))
    }

    fun executeFunction(virtualFrame: VirtualFrame): WaffleFunction {
        return WaffleTypesGen.expectWaffleFunction(execute(virtualFrame))
    }

    protected fun isArgumentIndexInRange(virtualFrame: VirtualFrame, index: Int): Boolean {
        return index < virtualFrame.arguments.size - 1
    }

    protected fun getArgument(virtualFrame: VirtualFrame, index: Int): Any {
        return virtualFrame.arguments[index + 1]
    }
}
